File: src/providers/tiptap/rules/python/script_src_hardcoded_api_key.py
"""tiptap-script-src-hardcoded-api-key (python analog, security)

Tiptap itself has no Python SDK, but Python backends that server-render HTML
embedding a Tiptap-hosted widget (e.g. the desmos/calculator node-view CDN
script) build the same `<script src="...?apiKey=...">` URL pattern. This
mirrors the JS rule: flag a `src=` assignment/keyword/dict-value whose string
(plain or f-string) contains a hardcoded `apiKey=`/`api_key=` value instead of
reading it from `os.environ`/`os.getenv`.

Flags:
    script.src = "https://cdn.example.com/embed.js?apiKey=abc123"
    Script(src=f"https://cdn.example.com/embed.js?apiKey={API_KEY}")  # API_KEY is a literal
    config["src"] = "...?api_key=abc123"

Does NOT flag:
    script.src = f"https://cdn.example.com/embed.js?apiKey={os.environ['API_KEY']}"
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))

from utils import is_env_access, key_name, loc  # noqa: E402

RULE_KEY = "tiptap-script-src-hardcoded-api-key"

_KEY_MARKERS = ("apiKey=", "api_key=")


def _has_hardcoded_key(node: ast.AST | None) -> bool:
    """Mirrors the JS quasi-walk: text before the marker is enough evidence,
    unless the value immediately following the marker is an env lookup."""
    if node is None:
        return False

    if isinstance(node, ast.Constant) and isinstance(node.value, str):
        return any(marker in node.value for marker in _KEY_MARKERS)

    if isinstance(node, ast.JoinedStr):
        values = node.values
        for i, part in enumerate(values):
            if not (isinstance(part, ast.Constant) and isinstance(part.value, str)):
                continue
            if not any(marker in part.value for marker in _KEY_MARKERS):
                continue
            # Marker text found in this literal chunk — check whether the very
            # next interpolated expression is an env lookup.
            next_expr = values[i + 1] if i + 1 < len(values) else None
            if isinstance(next_expr, ast.FormattedValue):
                if is_env_access(next_expr.value):
                    continue
                return True
            # Marker is in a tail chunk (or immediately followed by another
            # literal chunk) with no env-access expression right after it.
            return True
        return False

    if isinstance(node, ast.BinOp) and isinstance(node.op, ast.Add):
        return _has_hardcoded_key(node.left) or _has_hardcoded_key(node.right)

    return False


def _is_src_target(target: ast.AST) -> bool:
    if isinstance(target, ast.Attribute) and target.attr == "src":
        return True
    if isinstance(target, ast.Subscript):
        sl = target.slice
        return isinstance(sl, ast.Constant) and sl.value == "src"
    return False


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []

    for node in ast.walk(tree):
        # obj.src = "..."
        if isinstance(node, ast.Assign):
            for target in node.targets:
                if _is_src_target(target) and _has_hardcoded_key(node.value):
                    line, col, end_line, end_col = loc(node)
                    out.append(
                        {
                            "file": path,
                            "line": line,
                            "column": col,
                            "endLine": end_line,
                            "endColumn": end_col,
                            "ruleKey": RULE_KEY,
                        }
                    )
        elif isinstance(node, ast.AnnAssign) and node.value is not None:
            if _is_src_target(node.target) and _has_hardcoded_key(node.value):
                line, col, end_line, end_col = loc(node)
                out.append(
                    {
                        "file": path,
                        "line": line,
                        "column": col,
                        "endLine": end_line,
                        "endColumn": end_col,
                        "ruleKey": RULE_KEY,
                    }
                )

        # {"src": "..."} dict literal (config-driven node views / SSR templates)
        elif isinstance(node, ast.Dict):
            for k, v in zip(node.keys, node.values):
                if key_name(k) == "src" and _has_hardcoded_key(v):
                    line, col, end_line, end_col = loc(v)
                    out.append(
                        {
                            "file": path,
                            "line": line,
                            "column": col,
                            "endLine": end_line,
                            "endColumn": end_col,
                            "ruleKey": RULE_KEY,
                        }
                    )

        # Script(src="...") / tag.script(src="...") keyword-argument HTML builders
        elif isinstance(node, ast.Call):
            for kw in node.keywords:
                if kw.arg == "src" and _has_hardcoded_key(kw.value):
                    line, col, end_line, end_col = loc(node)
                    out.append(
                        {
                            "file": path,
                            "line": line,
                            "column": col,
                            "endLine": end_line,
                            "endColumn": end_col,
                            "ruleKey": RULE_KEY,
                        }
                    )

    return out

File: src/providers/tiptap/rules/js/script-src-hardcoded-api-key.ts
/**
 * tiptap-script-src-hardcoded-api-key (security)
 *
 * Detects when script.src is assigned a string literal (plain or template)
 * that contains `apiKey=` with a hardcoded value instead of an env variable.
 *
 * Flags:
 *   script.src = "https://cdn.example.com/calc.js?apiKey=abc123";
 *   script.src = `https://cdn.example.com/calc.js?apiKey=${LITERAL}`;
 *
 * Does NOT flag:
 *   script.src = `https://cdn.example.com/calc.js?apiKey=${process.env.KEY}`;
 */

const rule = {
  meta: {
    type: 'problem',
    docs: {
      description: 'script.src must not contain a hardcoded API key',
      category: 'security',
      cwe: 'CWE-798',
      owasp: 'API8:2023 Security Misconfiguration',
      rationale:
        'Hardcoding an API key directly in a script URL string commits the credential to version control and ships it in the client bundle. The key is trivially visible to anyone who views the page source. Read the key from an environment variable at runtime so it can be rotated without a code change and kept out of the repository.',
      docsUrl: 'https://tiptap.dev/docs/editor/extensions/custom-extensions/node-views',
      recommended: true,
    },
    messages: {
      hardcodedApiKey:
        'script.src contains a hardcoded API key. Read the key from an environment variable instead.',
    },
    schema: [],
  },
  create(context: any) {
    function isSrcAssignment(node: any): boolean {
      return (
        node?.type === 'AssignmentExpression' &&
        node.left?.type === 'MemberExpression' &&
        node.left.property?.type === 'Identifier' &&
        node.left.property.name === 'src'
      );
    }

    function isEnvAccess(node: any): boolean {
      // process.env.SOMETHING or process.env['SOMETHING']
      if (node?.type !== 'MemberExpression') return false;
      const obj = node.object;
      return (
        obj?.type === 'MemberExpression' &&
        obj.object?.type === 'Identifier' &&
        obj.object.name === 'process' &&
        obj.property?.type === 'Identifier' &&
        obj.property.name === 'env'
      );
    }

    return {
      AssignmentExpression(node: any) {
        if (!isSrcAssignment(node)) return;
        const right = node.right;

        // Plain string literal with apiKey= in it
        if (right?.type === 'Literal' && typeof right.value === 'string') {
          if (right.value.includes('apiKey=')) {
            context.report({ node, messageId: 'hardcodedApiKey' });
          }
          return;
        }

        // Template literal: `...apiKey=...` where the value after apiKey= is a literal expression
        if (right?.type === 'TemplateLiteral') {
          const quasis: any[] = right.quasis ?? [];
          const expressions: any[] = right.expressions ?? [];

          for (let i = 0; i < quasis.length; i++) {
            const cooked: string = quasis[i]?.value?.cooked ?? '';
            if (cooked.includes('apiKey=')) {
              // The key follows this quasi — check if the next expression is env access
              const nextExpr = expressions[i];
              if (!nextExpr) {
                // apiKey= is in a tail quasi with no following expression → hardcoded
                context.report({ node, messageId: 'hardcodedApiKey' });
                return;
              }
              if (!isEnvAccess(nextExpr)) {
                // Expression after apiKey= is not process.env.* → hardcoded
                context.report({ node, messageId: 'hardcodedApiKey' });
                return;
              }
            }
          }
        }
      },
    };
  },
};

export const tiptapScriptSrcHardcodedApiKeyRule = rule;
export default rule;
